### PR TITLE
feat: QR pairing + per-device tokens + Companion dialog + self-healing mobile reconnect

### PR DESCRIPTION
**Pull request by DaniAkash**
## Summary

Ships the entire pairing UX end-to-end. Desktop grows a Companion button in the StatusBar → dialog with QR + paired-devices list + Revoke. Mobile grows an `expo-camera` pair screen and, once paired, cold-starts through a self-healing resolver ladder rather than a manual URL+token form.

Design decisions were captured in the plan revision: mDNS `.local` in the QR, standardised port trio (`47823 / 28617 / 39482`), Devices UI absorbed into the pairing dialog, force-disconnect Revoke, resolver ladder scoped to same-LAN (cross-network deferred).

## What's inside

**Rust foundation** (`fa338a6`)
- `pairing.rs`: PairedTokens (SHA-256 device tokens in Keychain via `keyring 4`), expanded PairedDevice (id, platform, model, paired_at, last_seen, last_ip), PairingWindow with 5-min TTL + constant-time consume.
- `net.rs`: `.local` hostname via `hostname`, RFC 1918 IPv4 enumeration, `bind_first_available` walking the standard port trio.
- `mdns.rs`: `_agent-terminal._tcp.local.` DNS-SD advertisement via `mdns-sd`, fingerprint prefix in the TXT record.
- 12 new unit tests.

**Protocol + wss_server + Tauri commands** (`242c56c`)
- ClientFrame::PairingStart + ServerFrame::PairingComplete + bodies. Regenerated `protocol.gen.ts`.
- Auth branches: `PAIRING:<token>` prefix routes into a dedicated `pairing_flow` sub-task (60s timeout on the follow-up frame); paired-token match tags the connection with device_id and registers an abort oneshot in PairedConnMap; legacy dev bearer still available during the migration window.
- Revoke calls into PairedConnMap.revoke_and_close → forces AuthFail(revoked) + socket close within milliseconds AND blocks future auth with the revoked token.
- WSS bootstrap uses the port trio, publishes mDNS, exposes BoundNet + PairedTokens + PairingWindow as managed state.
- New Tauri commands: `open_pairing_window`, `close_pairing_window`, `get_pairing_qr_payload`, `list_paired_devices`, `revoke_paired_device`.
- 4 new integration tests: round-trip, bad token, no-open-window, revoke force-close.

**Desktop UI** (`647f2a9`)
- `<CompanionButton />` next to `<ThemeToggle />` in StatusBarLeft (Lucide Smartphone icon).
- `<CompanionDialog />`: pairing QR + fingerprint (monospace, for visual compare) + paired-devices list with Revoke. QR renders via `qrcode.react`.
- shadcn AlertDialog for Revoke confirm; `pairing:complete` event listener refreshes both the QR (fresh pairing token so the user can immediately pair another device) and the devices list.

**Mobile companion** (`afea65e`)
- `$device` store persists { token, deviceId, host, ips, port, fingerprint, deviceHint, lastIp, lastPort } as one secure-store blob.
- `resolver.ts`: candidate ladder — cached IP+port → `.local`+cached port → `.local`+each standard port → IP fallback. 12 unit tests covering every rung + label helper.
- `client.autoConnect(device)`: 2s per-rung TCP probe, first success wins, resolved endpoint written back to $device on AuthOk.
- PairScreen with expo-camera scanner → visual fingerprint confirm → PairingStart round-trip → persisted DeviceRecord → `/` redirect. State-machine split across `PairScreen.tsx` + `pair.components.tsx`; pure helpers extracted to `pair.helpers.ts` with 19 unit tests.
- Home screen: unpaired state now offers "Scan QR to pair" primary CTA + "Use manual token (advanced)" as a fallback link.
- App boot: `_layout.tsx` loads DeviceRecord from secure-store and kicks the resolver ladder if present.

## Testing

Rust: 137 total green — 123 lib + 9 hook + 6 wss + 4 wss-pairing + 1 wss-tls.
Companion: fmt + lint + tsc + fallow all clean, 31 unit tests green (12 resolver + 19 pair.helpers).
Desktop: tsc + biome clean.

## Manual smoke plan

- [ ] Fresh install + first pair: factory-reset phone (or `expo-secure-store` clear), tap Companion on desktop, scan QR on phone, fingerprint visible on both surfaces, confirm on phone, land on desktop's Projects view with a live session.
- [ ] Restart phone → auto-connect via resolver ladder rung 1 (cached IP+port) in <500ms.
- [ ] Move phone to a different Wi-Fi and back → auto-connect via rung 2 (`.local`+cached port).
- [ ] Reboot desktop so it lands on a different port → auto-connect via rung 3 (`.local`+standard port fallback).
- [ ] Revoke device from desktop Companion dialog → phone shows AuthFail(revoked) and can't reconnect with the old token.
- [ ] Pair a second phone without disturbing the first.

## Follow-ups (out of scope, tracked in the Phase 2 plan)

- Rate limiting on auth failures (PR C).
- Rotating audit log (PR C).
- Mobile Settings "Unpair from desktop" (PR C).
- Cross-network reachability (deferred; Tailscale sidecar or hosted relay, separate planning task).
- Active mDNS service scan on the mobile side (needs a dev-client build; PR D).

### CODE REVIEW

File: src-tauri/src/pairing.rs [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+    pub fn check_and_touch(&self, token: &str, source_ip: Option<String>) -> Option<PairedDevice> {
+        let candidate_hash = sha256_hex(token);
+        let mut map = self.inner.lock().expect("paired_tokens lock poisoned");
+        let mut hit: Option<String> = None;
+        for (id, device) in map.iter() {
+            if constant_time_eq(device.token_hash.as_bytes(), candidate_hash.as_bytes()) {
+                hit = Some(id.clone());
+                break;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (DaniAkash) [reply src-tauri/src/pairing.rs:181]: Fixed in b0af4d7 — dropped the early `break` and let the loop iterate the full map. Individual comparisons stay byte-level constant-time via `constant_time_eq`; total runtime is now independent of match position so the docstring promise is honoured.

File: src-tauri/src/commands.rs [hunk 2]
```diff
@@ -283,6 +285,173 @@ pub async fn list_projects() -> Result<serde_json::Value, String> {
...
+) -> Result<(), String> {
+    window.close();
+    Ok(())
+}
+
+/// Snapshot the pairing QR payload without opening a fresh session.
+/// Used when the desktop UI re-mounts and wants to redisplay whatever
+/// pairing token is currently live server-side (e.g. React strict-mode
+/// double-mount). Returns None on the payload's `pairing_token` field
+/// if no session is currently open.
```
[the remark this answers is not in the document]
Comment (DaniAkash) [reply src-tauri/src/commands.rs]: Fixed in f9da6f2 — rewrote the docstring to describe the actual `Result<Option<PairingQrPayload>, _>` shape and to name all three conditions that produce `Ok(None)` (no open session, TLS disabled, or WSS bind failed at startup).

File: src-tauri/src/net.rs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        // Any real host returns something. Test environments in CI
+        // sometimes hand back a strange short name; both are fine as
+        // long as they end in `.local`.
+        let host = local_hostname().expect("hostname should resolve");
+        assert!(
+            host.ends_with(".local"),
+            "expected .local suffix, got {host}"
+        );
+        assert_eq!(host, host.to_lowercase(), "expected lowercase");
+    }
```
[the remark this answers is not in the document]
Comment (DaniAkash) [reply src-tauri/src/net.rs:126]: Fixed in b0af4d7 — test now mirrors the prod code, degrading to a `return` + stderr note when `local_hostname()` yields `None` (as it can on sandboxed CI with an unset or non-UTF8 hostname). Positive-path assertions still run for any real host.

File: companion/src/modules/stores/$device.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+export async function loadDeviceFromSecureStore(): Promise<void> {
+  try {
+    const raw = await SecureStore.getItemAsync(STORE_KEY)
+    if (!raw) {
+      $device.set({ loaded: true, record: null })
+      return
+    }
+    const record: DeviceRecord = JSON.parse(raw)
+    $device.set({ loaded: true, record })
+  } catch {
```
[the remark this answers is not in the document]
Comment (DaniAkash) [reply companion/src/modules/stores/$device.ts]: Fixed in b0af4d7 — added a runtime `isDeviceRecord` type-guard called from `loadDeviceFromSecureStore` before the parsed blob is trusted. Every field the resolver + client touches is validated (token/deviceId non-empty strings, port finite number, ips string array, etc.). Split into `$device.helpers.ts` + `$device.types.ts` so bun test can exercise the pure predicate without pulling react-native into its module graph; 8 unit tests cover accept + reject + edge-case paths.

File: src/components/Companion/CompanionDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+              </span>
+            </div>
+            <DevicesList
+              devices={devices}
+              onRevoke={(id) => setConfirmRevokeId(id)}
+            />
+          </div>
+
+          <DialogFooter>
+            <Button variant="secondary" onClick={() => closeCompanionDialog()}>
```
[the remark this answers is not in the document]
Comment (DaniAkash) [reply src/components/Companion/CompanionDialog.tsx]: Fixed in f9da6f2 — Done button now wraps the call in `void closeCompanionDialog()` so the returned promise is explicitly discarded, matching the pattern already used elsewhere in the same component.

File: src/components/Companion/CompanionDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+              Cancel
+            </AlertDialogCancel>
+            <AlertDialogAction
+              onClick={async () => {
+                const target = confirmRevokeId
+                setConfirmRevokeId(null)
+                if (target === null) return
+                try {
+                  await revokeDevice(target)
+                } catch (_e) {}
```
[the remark this answers is not in the document]
Comment (DaniAkash) [reply src/components/Companion/CompanionDialog.tsx]: Fixed in 047ed47 — revoke failures now go into `$companion.error` via `$companion.setKey("error", ...)`, using the same error rendering surface the pairing section already had ("Revoke failed: <reason>"). Success path clears the error so a stale message from a previous action does not linger.

File: companion/src/modules/wss/client.ts [hunk 7]
```diff
@@ -71,6 +71,82 @@ export function connect(url: string, token: string): void {
...
+ */
+export async function autoConnect(
+  device: import('@/modules/stores/$device').DeviceRecord,
+): Promise<void> {
+  const { resolveWssCandidates } = await import('./resolver')
+  const candidates = resolveWssCandidates(device)
+  console.log('[wss.client] autoConnect ladder', {
+    candidates,
+    hint: device.deviceHint,
+  })
```
[the remark this answers is not in the document]
Comment (DaniAkash) [reply companion/src/modules/wss/client.ts:96]: Fixed in 047ed47 — up-front log now emits `{ rungs, hint }` (count + human hint only, no IPs or `.local` names), and the winning attempt logs `{ rung }` so debugging still knows which strategy resolved without exposing the addresses themselves.

File: src-tauri/src/wss_server.rs [hunk 8]
```diff
@@ -369,17 +503,142 @@ async fn connection_task(mut socket: WebSocket, state: Arc<ServerState>) {
...
+        Ok(Ok(None)) => return, // client closed
+        Ok(Err(e)) => {
+            eprintln!("[wss] pairing read error: {e}");
+            return;
+        }
+        Err(_) => {
+            let _ = send_frame(
+                &mut socket,
+                &ServerFrame::AuthFail {
+                    reason: "pairing_start not received within 60s".to_string(),
```
[the remark this answers is not in the document]
Comment (DaniAkash) [reply src-tauri/src/wss_server.rs]: Fixed in b0af4d7 — server timeout reason now starts with "pairing_start timeout: …" so the mobile `mapPairErrorFromException` substring-match routes to the dedicated `timeout` UI state instead of the generic connection-failed fallback. Own-goal from the same PR; both sides now agree.

File: src/components/Companion/CompanionDialog.tsx [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+  useEffect(() => {
+    if (!open) return
+    let unlisten: (() => void) | undefined
+    void listenForPairingComplete().then((fn) => {
+      unlisten = fn
+    })
+    return () => {
+      unlisten?.()
+    }
+  }, [open])
```
[the remark this answers is not in the document]
Comment (DaniAkash) [reply src/components/Companion/CompanionDialog.tsx:68]: Fixed in af28f50 — added a `cancelled` flag to the effect. If cleanup fires before `listenForPairingComplete()` resolves, the pending `.then()` sees `cancelled = true` and disposes the freshly-returned listener immediately; otherwise cleanup calls the now-set `unlisten`. No leaked subscriptions across open/close cycles.

File: src-tauri/src/commands.rs [hunk 10]
```diff
@@ -283,6 +285,178 @@ pub async fn list_projects() -> Result<serde_json::Value, String> {
...
+    /// Human label the mobile shows during pairing so the user knows
+    /// which desktop they're pairing with. Derived from the hostname
+    /// (title-cased, `.local` stripped).
+    pub device_hint: String,
+}
+
+/// Mint a fresh pairing token and return the full QR payload the
+/// desktop UI encodes into the on-screen QR. Fails when TLS is off
+/// (no fingerprint to embed) or when the WSS bind failed at startup
+/// (nothing for the mobile to connect to).
```
[the remark this answers is not in the document]
Comment (DaniAkash) [reply src-tauri/src/commands.rs]: Fixed in 3311975 — rewrote the docstring to describe the actual behavior. TLS-off is no longer a failure condition; it emits an empty-string fingerprint sentinel that both the desktop dialog and the mobile pair screen render as a "TLS off (dev only)" banner. The only failure is a WSS bind that failed at startup.

File: src-tauri/src/commands.rs [hunk 11]
```diff
@@ -283,6 +285,178 @@ pub async fn list_projects() -> Result<serde_json::Value, String> {
...
+    Ok(())
+}
+
+/// Snapshot the pairing QR payload without opening a fresh session.
+/// Used when the desktop UI re-mounts and wants to redisplay whatever
+/// pairing token is currently live server-side (e.g. React strict-mode
+/// double-mount). Returns `Ok(None)` when no session is currently
+/// open, TLS is disabled, or the WSS bind failed at startup; the
+/// caller should treat that as "nothing to display, tell the user to
+/// open a new pairing session".
```
[the remark this answers is not in the document]
Comment (DaniAkash) [reply src-tauri/src/commands.rs]: Fixed in 3311975 — same drift, same fix. Docstring now clarifies that `Ok(None)` means "no pairing session open OR WSS bind failed", and that TLS-off yields `Some(payload)` with an empty-string fingerprint (paralleling `open_pairing_window`).

File: src-tauri/src/wss_server.rs [hunk 12]
```diff
@@ -268,23 +374,36 @@ async fn connection_task(mut socket: WebSocket, state: Arc<ServerState>) {
...
+    // Step 2: dispatch to the right auth branch.
+    if let Some(pairing_token) = token.strip_prefix(PAIRING_PREFIX) {
+        pairing_flow(socket, &state, pairing_token).await;
         return;
     }
 
+    let (session_device_name, device_id) =
+        match state.paired_tokens.check_and_touch(&token, None) {
+            Some(dev) => (dev.device_name.clone(), Some(dev.id.clone())),
+            None => {
```
[the remark this answers is not in the document]
Comment (DaniAkash) [reply src-tauri/src/wss_server.rs:395]: Fixed in af28f50 — threaded the peer `SocketAddr` from `handle_stream_upgrade` through `connection_task` and into `check_and_touch(&token, Some(peer.ip().to_string()))`. `PairedDevice.last_ip` now populates on every successful paired-token auth, matching the field docstring and the Companion dialogs troubleshooting surface.

File: src-tauri/src/wss_server.rs [hunk 13]
```diff
@@ -369,17 +518,164 @@ async fn connection_task(mut socket: WebSocket, state: Arc<ServerState>) {
...
         state.hub.unsubscribe(&tab_id, sub_id);
     }
 }
 
+/// Handle the `PAIRING:` sub-flow. The auth path already validated the
+/// pairing token via `PairingWindow::consume`. Here we send AuthOk
+/// with a "Pairing" placeholder, wait for exactly one PairingStart
+/// frame (60s timeout so a phone that camera-scans and never taps the
+/// confirm button doesn't leak a connection), mint the long-lived
+/// device token, send PairingComplete, then close.
```
[the remark this answers is not in the document]
Comment (DaniAkash) [reply src-tauri/src/wss_server.rs]: Fixed in 3311975 — rewrote the doc to describe the validate-then-consume ordering that has been in place since the "dont burn the pair token when insert fails" commit. Explains the reason (transient Keychain refusal on first-run macOS prompt should not strand the user with a dead pair window) so a future change knows why the two-step is deliberate.
